Adjust <input type=checkbox switch> iOS animation

annevk opened the pull request:
#### fb6f64f9959f2873795185a5be1348f9c2d75cbb
<pre>
Adjust &lt;input type=checkbox switch&gt; iOS animation
<a href="https://bugs.webkit.org/show_bug.cgi?id=271839">https://bugs.webkit.org/show_bug.cgi?id=271839</a>
<a href="https://rdar.apple.com/125563501">rdar://125563501</a>

Reviewed by Aditya Keerthi.

The pressed animation for the thumb needs a slight delay. This in turn
necessitates introducing additional state to indicate when it is
pressed, as the control generally being active is no longer sufficient
(that changes too early now).

Also make these more general improvements:

- Rename pressed to held to account for the slight change in semantics.
- Abstract finding the relevant Touch of a TouchEvent in a method.
- Call dispatchSimulatedClick with SendNoEvents for its mouse events
  argument. They are redundant and could possibly lead to non-touch
  code paths being triggered.

* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::findTouchWithIdentifier):
(WebCore::CheckboxInputType::subsequentTouchEventTouch const):
(WebCore::CheckboxInputType::handleTouchEvent):
(WebCore::CheckboxInputType::startSwitchPointerTracking):
(WebCore::CheckboxInputType::stopSwitchPointerTracking):
(WebCore::CheckboxInputType::disabledStateChanged):
(WebCore::CheckboxInputType::willUpdateCheckedness):
(WebCore::switchAnimationDuration):
(WebCore::CheckboxInputType::switchAnimationStartTime const):
(WebCore::CheckboxInputType::setSwitchAnimationStartTime):
(WebCore::CheckboxInputType::setIsSwitchHeld):
(WebCore::CheckboxInputType::isSwitchVisuallyOn const):
(WebCore::CheckboxInputType::switchAnimationHeldProgress const):
(WebCore::CheckboxInputType::isSwitchHeld const):
(WebCore::CheckboxInputType::switchAnimationTimerFired):
(WebCore::CheckboxInputType::switchAnimationPressedProgress const): Deleted.
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::switchAnimationHeldProgress const):
(WebCore::HTMLInputElement::isSwitchHeld const):
(WebCore::HTMLInputElement::switchAnimationPressedProgress const): Deleted.
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/platform/graphics/SpringSolver.h:
(WebCore::SpringSolver::solve const):
(WebCore::SpringSolver::solve): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::switchAnimationHeldDuration const):
(WebCore::RenderTheme::switchAnimationPressedDuration const): Deleted.
* Source/WebCore/rendering/ios/RenderThemeIOS.h:

Canonical link: <a href="https://commits.webkit.org/278311@main">https://commits.webkit.org/278311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d860986614d1694472ac2d7d8e0cee81b55f551

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2462 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53427 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/859 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52472 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35686 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/425 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40939 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27134 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43176 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22036 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24557 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/427 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8547 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46545 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/488 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55010 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25265 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/390 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48333 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26526 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43361 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47346 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27389 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7246 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26258 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -154,45 +165,44 @@ void CheckboxInputType::handleTouchEvent(TouchEvent& event)
+
+        m_switchPointerTrackingTouchIdentifier = touch->identifier();
+        if (!m_switchPressedTimer) {
+            m_switchPressedTimer = makeUnique<Timer>([this, touch] {
```
This should capture a `WeakPtr`.

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2373,4 +2373,10 @@ float HTMLInputElement::switchAnimationPressedProgress() const
     return downcast<CheckboxInputType>(*m_inputType).switchAnimationPressedProgress();
 }
 
+bool HTMLInputElement::isSwitchPressed() const
```
I don't think this method is called anywhere?

annevk replied:
This is needed by the internal drawing code.

I was thinking that maybe we want to put this behind IOS_TOUCH_EVENTS as well for now to reduce impact on other platforms.

pxlcoder replied:
I think we should guard it somehow, otherwise there's nothing stopping someone from accidentally removing it.

pxlcoder replied:
I also wonder if this should be `isSwitchHeld`? The switch is technically pressed before the timer fires.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -159,45 +170,44 @@ void CheckboxInputType::handleTouchEvent(TouchEvent& event)
+            m_switchHeldTimer->stop();
+        if (m_isSwitchHeld)
+            performSwitchAnimation(SwitchAnimationType::Held);
+        m_isSwitchHeld = false;
```
Good use case for `std::exchange`.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -159,45 +170,44 @@ void CheckboxInputType::handleTouchEvent(TouchEvent& event)
+                protectedThis->setIsSwitchHeld(true);
+            });
+        }
+        m_switchHeldTimer->startOneShot(200_ms);
```
`200_ms` should probably be in a `static constexpr` variable near the top of this file.

Similar to `static constexpr unsigned maxSelectItems = 100000;` in `HTMLSelectElement.cpp`.

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -133,19 +133,30 @@ void CheckboxInputType::handleMouseMoveEvent(MouseEvent& event)
     return nullptr;
 }
 
+Touch* CheckboxInputType::subsequentTouchEventTouch(TouchEvent& event) const
```
Can the parameter be `const TouchEvent& event`?

annevk committed:
Adjust <input type=checkbox switch> iOS animation
https://bugs.webkit.org/show_bug.cgi?id=271839
rdar://125563501

Reviewed by Aditya Keerthi.

The pressed animation for the thumb needs a slight delay. This in turn
necessitates introducing additional state to indicate when it is
pressed, as the control generally being active is no longer sufficient
(that changes too early now).

Also make these more general improvements:

- Rename pressed to held to account for the slight change in semantics.
- Abstract finding the relevant Touch of a TouchEvent in a method.
- Call dispatchSimulatedClick with SendNoEvents for its mouse events
  argument. They are redundant and could possibly lead to non-touch
  code paths being triggered.

* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::findTouchWithIdentifier):
(WebCore::CheckboxInputType::subsequentTouchEventTouch const):
(WebCore::CheckboxInputType::handleTouchEvent):
(WebCore::CheckboxInputType::startSwitchPointerTracking):
(WebCore::CheckboxInputType::stopSwitchPointerTracking):
(WebCore::CheckboxInputType::disabledStateChanged):
(WebCore::CheckboxInputType::willUpdateCheckedness):
(WebCore::switchAnimationDuration):
(WebCore::CheckboxInputType::switchAnimationStartTime const):
(WebCore::CheckboxInputType::setSwitchAnimationStartTime):
(WebCore::CheckboxInputType::setIsSwitchHeld):
(WebCore::CheckboxInputType::isSwitchVisuallyOn const):
(WebCore::CheckboxInputType::switchAnimationHeldProgress const):
(WebCore::CheckboxInputType::isSwitchHeld const):
(WebCore::CheckboxInputType::switchAnimationTimerFired):
(WebCore::CheckboxInputType::switchAnimationPressedProgress const): Deleted.
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::switchAnimationHeldProgress const):
(WebCore::HTMLInputElement::isSwitchHeld const):
(WebCore::HTMLInputElement::switchAnimationPressedProgress const): Deleted.
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/platform/graphics/SpringSolver.h:
(WebCore::SpringSolver::solve const):
(WebCore::SpringSolver::solve): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::switchAnimationHeldDuration const):
(WebCore::RenderTheme::switchAnimationPressedDuration const): Deleted.
* Source/WebCore/rendering/ios/RenderThemeIOS.h:

Canonical link: https://commits.webkit.org/278311@main
